src/puppetlabs/puppetdb/scf/storage.clj

```diff
@@ -1433,6 +1433,79 @@
+   pdb database which have granted locks with the exception of those queries
+   from the gc-pid or the killer's pid. This should clear the way for GC to
+   grab the lock it's requesting in the main GC thread. This loop repeats
+   until the killer thread is interrupted or receives confirmation from the
+   GC thread via the gc-finished? atom that the GC thread has dropped the
+   partition."
+  [db gc-pid gc-finished?]
+  (let [{:keys [current_database current_user]}
+        (first (jdbc/query-to-vec "select current_database(), current_user;"))
+        kill-blocking-qs (format
+                          "select pg_cancel_backend(l.pid)
+                            from pg_locks l inner join pg_stat_activity a on l.pid = a.pid
+                            where (a.datname = '%s' AND a.usename = '%s')
+                             and not (l.pid = pg_backend_pid() OR l.pid = %s)
+                             and l.granted = true
+                             and exists (select 1
+                                          from pg_locks
+                                          where pid = %s and mode = 'AccessExclusiveLock'
+                                           and granted = false)
+                            group by l.pid;"
+                          current_database current_user gc-pid gc-pid)]
+    (Thread.
+     #(with-noisy-failure
+        (try
+          ;; you can't use *db* value from the gc thread because the connection
+          ;; it has may already be blocked attempting to drop the partition.
+          ;; Binding *db* to the passed in db var will cause the killer thread
+          ;; to get a new Hikari connection
+          (binding [jdbc/*db* db]
+            ;; Q: is there any possible way that all queries could be in a blocked
+            ;; state where we're unable to get another connection from the pool
+            ;; to issue the kill command?
+            (while (not @gc-finished?)
+              (jdbc/do-commands kill-blocking-qs)
+              (Thread/sleep 1000)))
+          ;; Q: do we need to worry about other exceptions here as well?
+          (catch InterruptedException ex
+            ;; the killer thread may be interrupted by the GC thread if it's
+            ;; in the sleep or do-commands part of the while loop
+            true))))))
+
+(def gc-query-killer-timeout-ms
+  (env-config-for-db-ulong "PDB_GC_QUERY_KILLER_TIMEOUT_MS"
+                           (* 20 60 1000)))
+
+(defn maybe-launch-query-killer [drop-one candidate db]
+  (if (pos? gc-query-killer-timeout-ms)
+    (let [gc-pid (-> "select pg_backend_pid();"
+                     jdbc/query-to-vec
+                     first
+                     :pg_backend_pid)
+          gc-finished? (atom false)
+          query-killer (spawn-query-killer db gc-pid gc-finished?)]
+      (try
+        (.start query-killer)
+        (drop-one candidate)
+        (reset! gc-finished? true)
```

Comment:
Moving the `gc-finished?` reset into the finally block makes sense to me. One advantage of keeping the interrupt would be that it can break us out of the sleep in the killer thread's loop right away which could mean a shorter wait on the `.join` call. That's a good point about the `jdbc` stuff possibly not supporting thread interruption. We might want to add a timeout directly to the query in this case and double check what we're doing in `sync` wrt thread interruption.